packages/flutter/lib/src/widgets/autocomplete.dart

```diff
@@ -526,6 +640,87 @@ class _RawAutocompleteState<T extends Object> extends State<RawAutocomplete<T>>
   }
 }
 
+/// Positions the options view.
+class _OptionsLayoutDelegate extends SingleChildLayoutDelegate {
+  _OptionsLayoutDelegate({
+    required this.fieldSize,
+    required this.fieldOffset,
+    required this.optionsViewOpenDirection,
+    required TextDirection? textDirection,
+  }) : textDirection = textDirection ?? TextDirection.ltr;
+
+  /// The size of the field in [RawAutocomplete.fieldViewBuilder].
+  final Size? fieldSize;
+
+  /// The position of the field in [RawAutocomplete.fieldViewBuilder].
+  final Offset? fieldOffset;
+
+   /// A direction in which to open the options view overlay.
+  final OptionsViewOpenDirection optionsViewOpenDirection;
+
+  /// The [TextDirection] of this part of the widget tree.
+  final TextDirection textDirection;
+
+  // Limits the child to the space above/below the field, with a minimum, and
+  // with the same maxWidth constraint as the field has.
+  @override
+  BoxConstraints getConstraintsForChild(BoxConstraints constraints) {
+    // shouldRelayout prevents this from being called if fieldOffset or
+    // fieldSize are invalid.
+    assert(fieldSize != null && fieldSize!.isFinite);
+    assert(fieldOffset != null && fieldOffset!.isFinite);
+
+    return constraints.loosen().copyWith(
+      // The field width may be zero if this is a split RawAutocomplete with no
+      // field of its own. In that case, don't change the constraints width.
+      maxWidth: fieldSize!.width == 0.0 ? constraints.maxWidth : fieldSize!.width,
+      maxHeight: switch (optionsViewOpenDirection) {
+        OptionsViewOpenDirection.down => max(
+          min(kMinInteractiveDimension, constraints.maxHeight),
+          constraints.maxHeight - fieldOffset!.dy - fieldSize!.height,
+        ),
+        OptionsViewOpenDirection.up => max(kMinInteractiveDimension, fieldOffset!.dy),
+      },
```

Comment:
> Or can the region that this overlay is operating in be potentially only part of the space that the global coordinate system covers?

I'm thinking of a case where this might be true, but nothing comes to mind.

> If the latter is possible, I think we should be able to fix it by passing an appropriate ancestor parameter to localToGlobal.

Is it okay to use the Overlay renderbox as the ancestor? Since that is the only common ancestor of the field and options view that we can access?

Reply:
> to use the Overlay renderbox as the ancestor? Since that is the only common ancestor of the field and options view that we can access?

Using a known common ancestor like that sounds great.

The offset we get will then be relative to the Overlay's coordinate system. So we're assuming that that's the same coordinate system this render box is working in, modulo any adjustments we make explicitly like `verticalOffset`. That sounds right but I haven't looked into Overlay and OverlayPortal enough to be confident of it.


> Or can the region that this overlay is operating in be potentially only part of the space that the global coordinate system covers?

It looks like the one place within the framework that constructs an Overlay is in the Navigator build method.

So I think this sort of situation can happen if there's a Navigator that doesn't occupy the whole screen (or more precisely, whole `View`). This could happen if for some reason someone makes a WidgetsApp that's only part of the screen. There are also a couple of other spots that make a Navigator, like TabView, and it seems entirely reasonable that someone could have a TabView that's only part of the screen; or someone could use Navigator for a custom thing and have it not occupy the whole screen.